From 911b79cde95c7da0ec02f48105358a36636b7a71 Mon Sep 17 00:00:00 2001
From: David Howells <dhowells@redhat.com>
Date: Mon, 19 Oct 2015 11:20:28 +0100
Subject: [PATCH] KEYS: Don't permit request_key() to construct a new keyring
Git-commit: 911b79cde95c7da0ec02f48105358a36636b7a71
Patch-mainline: v4.3-rc7
References: CVE-2015-7872 bsc#951440
For-stable: no, depends on 4bdf0bc300314141e5475e145acb8b5ad846f00d

If request_key() is used to find a keyring, only do the search part - don't
do the construction part if the keyring was not found by the search.  We
don't really want keyrings in the negative instantiated state since the
rejected/negative instantiation error value in the payload is unioned with
keyring metadata.

Now the kernel gives an error:

	request_key("keyring", "#selinux,bdekeyring", "keyring", KEY_SPEC_USER_SESSION_KEYRING) = -1 EPERM (Operation not permitted)

Signed-off-by: David Howells <dhowells@redhat.com>
Acked-by: Lee, Chun-Yi <jlee@suse.com>
---
 security/keys/request_key.c |    3 +++
 1 file changed, 3 insertions(+)

--- a/security/keys/request_key.c
+++ b/security/keys/request_key.c
@@ -458,6 +458,9 @@ static struct key *construct_key_and_lin
 
 	kenter("");
 
+	if (ctx->index_key.type == &key_type_keyring)
+		return ERR_PTR(-EPERM);
+	
 	user = key_user_lookup(current_fsuid());
 	if (!user)
 		return ERR_PTR(-ENOMEM);
